django/contrib/auth/backends.py

```diff
@@ -57,32 +60,41 @@ def _get_permissions(self, user_obj, obj, from_name):
             setattr(user_obj, perm_cache_name, {"%s.%s" % (ct, name) for ct, name in perms})
         return getattr(user_obj, perm_cache_name)
 
-    def get_user_permissions(self, user_obj, obj=None):
+    def get_user_permissions(self, user_obj, obj=None, fallback_to_model=None):
         """
         Return a set of permission strings the user `user_obj` has from their
         `user_permissions`.
         """
-        return self._get_permissions(user_obj, obj, 'user')
+        return self._get_permissions(user_obj, obj, 'user', fallback_to_model)
 
-    def get_group_permissions(self, user_obj, obj=None):
+    def get_group_permissions(self, user_obj, obj=None, fallback_to_model=None):
         """
         Return a set of permission strings the user `user_obj` has from the
         groups they belong.
         """
-        return self._get_permissions(user_obj, obj, 'group')
+        return self._get_permissions(user_obj, obj, 'group', fallback_to_model)
 
-    def get_all_permissions(self, user_obj, obj=None):
-        if not user_obj.is_active or user_obj.is_anonymous or obj is not None:
+    def _get_fallback_setting(self, fallback_to_model):
+        if fallback_to_model is not None:
+            return fallback_to_model
+        return settings.OBJECT_PERMISSION_FALLBACK_TO_MODEL
+
+    def get_all_permissions(self, user_obj, obj=None, fallback_to_model=None):
+        fallback_to_model = self._get_fallback_setting(fallback_to_model)
+        if not user_obj.is_active or user_obj.is_anonymous \
+                or obj is not None and fallback_to_model is False:
             return set()
+
         if not hasattr(user_obj, '_perm_cache'):
             user_obj._perm_cache = {
                 *self.get_user_permissions(user_obj),
                 *self.get_group_permissions(user_obj),
             }
         return user_obj._perm_cache
```

Comment:
Using the cache like this here seems dangerous since you do not know how it got populated (ie was `fallback_to_model` true or false at the time of the population?)

Reply:
Actually it isn't. `fallback_to_model` acts like an on-off switch. Give it all, or none. Cache holds "all", and if not prevented by a `False` fallback value, provides it. 

Your concern, on the other hand, holds true for the backends that do implement object permissions. One is guardian, and the patch I sent them do take that into consideration. 

As far as the django's default backend is concerned, I have included plenty of tests. For example, global setting False: asking for model permissions, then object, then overriding the setting to True and asking for object again. Same for global setting of True. Every possible combination is tested.